### ISSUE TITLE
FocusWorld is unpredictable

### ISSUE THREAD

**Issue by ProbablePrime** [Frooxius/NeosPublic#1246]
Im trying to use the FocusWorld LogiX, node for the worlds you're in facet: https://youtu.be/Ugw7ffoUUj8

Due to changes Froox added as a result of #1235 WorldThumbnailItem now returns session ids when configured as per Froox's changes. This is Great!

I take the session ids out of the WorldThumbnailItem and add the prefix "neos-session:///". Then i plug that into FocusWorld and hook in a button click handler.

The majority of the time, focusworld isn't actually doing anything though. So i re-created the exact scenario in isolation:
1. Open World A
1. Spawn a Focus World Node and hook in an Impulse Input
1. Open World B
1. Obtain World B's 'Session URL' (Starts with neos-session://)
1. Take World B's session URL to World A
1. Connect World B's session URL to World A's Focus Node
1. Fire the impulse
1. Nothing Happens or Rarely it works :S

Strangely, There are cases where the node neither outputs "Focused" or "Not Found". To me that sounds like an error might be happening internally so here's a log file
[ALFRED - 2020.10.30.1068 - 2020-10-31 01_05_27.log](https://github.com/Frooxius/NeosPublic/files/5469956/ALFRED.-.2020.10.30.1068.-.2020-10-31.01_05_27.log)

From a quick scan of the log file there is a FocusWorld stack trace:
```
02:13:52.579 ( 90 FPS)	Exception running asynchronous task:
System.AggregateException: One or more errors occurred. ---> System.ArgumentNullException: Value cannot be null.
Parameter name: source
  at System.Linq.Enumerable.Any[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x0000d] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__0 (FrooxEngine.World w) [0x0000d] in <3f0261784f5b4b9b85b655a8ae301b6d>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x0001c] in <3f0261784f5b4b9b85b655a8ae301b6d>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <3f0261784f5b4b9b85b655a8ae301b6d>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ArgumentNullException: Value cannot be null.
Parameter name: source
  at System.Linq.Enumerable.Any[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x0000d] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__0 (FrooxEngine.World w) [0x0000d] in <3f0261784f5b4b9b85b655a8ae301b6d>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x0001c] in <3f0261784f5b4b9b85b655a8ae301b6d>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <3f0261784f5b4b9b85b655a8ae301b6d>:0 <---


   at void BaseX.UniLog.Error(string message, bool stackTrace)
   at void FrooxEngine.CoroutineManager.CheckExceptions(Task task)
   at void System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at void System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at bool System.Threading.ThreadPoolWorkQueue.Dispatch()
   at bool System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

Please see the log file, for more details.

This has prompted some questions:
1. Does FocusWorld, not work with private/friendsonly/hidden worlds? I am the owner of the impulse and therefore i should have access to those sessions?
1. If this is an error but not a bug(as in it cannot be fixed) which can be caught, perhaps add an onfail output impulse. This will let me know if it Focuses, Doesn't Find or Fails.

A reminder here that the end goal here is to put this in a facet just in case that's relevant.

Thanks!

**Comment by Frooxius** [comment]
Should be fixed in 2020.11.2.1161! This is because when using neos-session URL, it needs to be resolved first to actual URL's for connection, but for private/hidden sessions this doesn't work, since Neos Cloud isn't aware of their actual URLs. 

I've made change which bypasses this for focusing a world and just focuses directly based on the SessionID extracted from the URL. Let me know if you run into more issues.

**Comment by ProbablePrime** [comment]
Thank you Froox so much :)

**Comment by ProbablePrime** [comment]
Im still having trouble with this, here's a video for some reproduction cases: https://youtu.be/ISk4n0mS7ZY

```
21:59:32.098 ( 90 FPS)	CloudInfo: SessionInfo. Id: S-83c6825a-1d28-4980-b904-679bf880665f, Name: Desert Meeting Space, URLs: lnl-nat:///S-83c6825a-1d28-4980-b904-679bf880665f, neos-steam://76561197987309564/2/S-83c6825a-1d28-4980-b904-679bf880665f
21:59:32.110 ( 90 FPS)	Exception running asynchronous task:
System.AggregateException: One or more errors occurred. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__1 (FrooxEngine.World w) [0x00006] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x00029] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__1 (FrooxEngine.World w) [0x00006] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x00029] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 <---


   at void BaseX.UniLog.Error(string message, bool stackTrace)
   at void FrooxEngine.CoroutineManager.CheckExceptions(Task task)
   at void System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at void System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at bool System.Threading.ThreadPoolWorkQueue.Dispatch()
   at bool System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
21:59:36.187 ( 90 FPS)	Resolving SessionID: s-83c6825a-1d28-4980-b904-679bf880665f
21:59:36.187 ( 90 FPS)	LocalInfo: 
21:59:36.251 ( 90 FPS)	CloudInfo: SessionInfo. Id: S-83c6825a-1d28-4980-b904-679bf880665f, Name: Desert Meeting Space, URLs: lnl-nat:///S-83c6825a-1d28-4980-b904-679bf880665f, neos-steam://76561197987309564/2/S-83c6825a-1d28-4980-b904-679bf880665f
21:59:36.255 ( 90 FPS)	Exception running asynchronous task:
System.AggregateException: One or more errors occurred. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__1 (FrooxEngine.World w) [0x00006] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x00029] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__1 (FrooxEngine.World w) [0x00006] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x00029] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 <---


   at void BaseX.UniLog.Error(string message, bool stackTrace)
   at void FrooxEngine.CoroutineManager.CheckExceptions(Task task)
   at void System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at void System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at bool System.Threading.ThreadPoolWorkQueue.Dispatch()
   at bool System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
22:00:00.366 ( 90 FPS)	Resolving SessionID: s-83c6825a-1d28-4980-b904-679bf880665f
22:00:00.366 ( 90 FPS)	LocalInfo: 
22:00:00.425 ( 90 FPS)	CloudInfo: SessionInfo. Id: S-83c6825a-1d28-4980-b904-679bf880665f, Name: Desert Meeting Space, URLs: lnl-nat:///S-83c6825a-1d28-4980-b904-679bf880665f, neos-steam://76561197987309564/2/S-83c6825a-1d28-4980-b904-679bf880665f
22:00:00.434 ( 90 FPS)	Exception running asynchronous task:
System.AggregateException: One or more errors occurred. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__1 (FrooxEngine.World w) [0x00006] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x00029] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.Userspace+<>c__DisplayClass181_0.<GetExistingWorld>b__1 (FrooxEngine.World w) [0x00006] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate, System.Boolean& found) [0x0003f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.FirstOrDefault[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at FrooxEngine.Userspace.GetExistingWorld (FrooxEngine.WorldStartSettings startInfo) [0x00029] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 
  at FrooxEngine.LogiX.WorldNodes.FocusWorld+<>c__DisplayClass5_0+<<Focus>b__0>d.MoveNext () [0x0009a] in <6bfa52c7566b4f9c8b0e0477ef177a83>:0 <---


   at void BaseX.UniLog.Error(string message, bool stackTrace)
   at void FrooxEngine.CoroutineManager.CheckExceptions(Task task)
   at void System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at void System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at bool System.Threading.ThreadPoolWorkQueue.Dispatch()
   at bool System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

Here's the log file: https://i.probableprime.co.uk/201103-ABwcjynh7YnSdMkWLwpH.log

**Comment by art0007i** [comment]
I have tried to modify the worlds you are in facet to have this functionality myself, and I also ran into this problem, from my research it seems like the neos-session:/// URL never works, if the host is using steam then the working URL will be neos-steam:///{STEAMID}/{SESSIONID}, alternatively lnl-nat:///{SESSIONID} seems to always work. The other problem I ran into is that it seems impossible to focus worlds using the logix node in the userspace.

**Comment by Frooxius** [comment]
Should be fixed in 2020.11.5.88
